### PR TITLE
Optimized implementation using topological sort

### LINKED ISSUE #43: Runtime performance
Running the critical path method on a "big" activity graph (200+ nodes, 10.000+ edges), performance can be a problem. In one of our use cases, the algorithm takes roughly 15 minutes on my machine, to finish.

### PR DESCRIPTION
**Pull request by NiklasHansen**
The "old" implementation was not really optimized, leading to poor performance on "big" activity graphs (200+ nodes, 10.000+ edges). This implementation does a topological sort first, and then does linear iterations over the nodes.

With my local test use-case, runtimes dropped from roughly 15m to 0.1s.

Fixes #43

### CODE REVIEW

**Comment by simon-k** [comment]
Looks good!

Would you agree that this also makes #11 somewhat redundant since performance is not an issue anymore?

File: src/NovoNordisk.CriticalPath/CriticalPathMethod.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public interface ICriticalPathMethod
     /// <param name="activities">List of activities. Note that the HashSet is call-by-reference, so its properties will be modified by this function.</param>
     /// <returns>The activities on the critical path, ordered by their dependencies.</returns>
     List<Activity> Execute(HashSet<Activity> activities);
-}
+}
```
Comment (simon-k) [inline src/NovoNordisk.CriticalPath/CriticalPathMethod.cs:16]: Spaces

File: src/NovoNordisk.CriticalPath/CriticalPathMethod.cs [hunk 2]
```diff
@@ -25,117 +25,128 @@ public class CriticalPathMethod : ICriticalPathMethod
...
+        
+        foreach (var activity in activities)
         {
-            criticalActivities.Add(nextActivity);
-            nextActivity = nextActivity.Dependencies.FirstOrDefault(_ => _.TotalFloat == 0);
+            AddEdgesFromActivity(activity, edges, visited);
         }
-
-        return criticalActivities;
+
```
Comment (simon-k) [inline src/NovoNordisk.CriticalPath/CriticalPathMethod.cs:47]: Spaces

File: src/NovoNordisk.CriticalPath/CriticalPathMethod.cs [hunk 3]
```diff
@@ -25,117 +25,128 @@ public class CriticalPathMethod : ICriticalPathMethod
...
-            initialActivity.EarlyStart = 0;
-            initialActivity.EarlyFinish = initialActivity.Cost;
-            SetEarly(initialActivity);
+            activity.LatestFinish = latestFinish;
+            activity.LatestStart = latestFinish - activity.Cost;
         }
     }
-
-    private static void SetEarly(Activity activity)
+
```
Comment (simon-k) [inline src/NovoNordisk.CriticalPath/CriticalPathMethod.cs:137]: Spaces

**Review by simon-k — APPROVED** [review]

**Comment by NiklasHansen** [comment]
> Looks good!
> 
> @NiklasHansen would you agree that this also makes #11 somewhat redundant since performance is not an issue anymore?

For performance reasons, then yes, I would regard that issue as redundant. And we do not have any I/O or similar, that would benefit from async/await either.

**Pull request merged**